### PR TITLE
dbt deps prerelease install bugs + add install-prerelease parameter to packages.yml

### LINKED ISSUE #3578: Package resolution is inconsistent for prerelease versions
### Describe the bug
The incorrect package is installed using package/version syntax, but the correct package is installed using git/revision syntax.

### Steps To Reproduce

Set packages.yml to install 0.7.6 version of dbtvault.
Run dbt deps.

The wrong version of dbtvault is installed. Below packages.yml is set to 0.7.6 but b1 variation is installed.

Same behaviour on 0.19.1 and 0.20.0

### Expected behavior
Expected that the package version listed is the one installed.

### Screenshots and log output
```
❯ cat packages.yml
packages:
  - package: Datavault-UK/dbtvault
    version: 0.7.6
❯ dbt deps
Running with dbt=0.19.1
No profile "default" found, continuing with no target
Installing Datavault-UK/dbtvault@0.7.6-b1
  Installed from version 0.7.6-b1

Installing fishtown-analytics/dbt_utils@0.6.4
  Installed from version 0.6.4
```
### System information
**Which database are you using dbt with?**
- [ ] postgres
- [ ] redshift
- [ ] bigquery
- [ ] snowflake
- [x] other (specify: Not applicable.)

**The output of `dbt --version`:**
```
❯ dbt --version
installed version: 0.20.0
   latest version: 0.20.0

Up to date!
```

**The operating system you're using:**
OSX, but also occurs in dbt Cloud.

**The output of `python --version`:**
```
❯ python --version
Python 2.7.16
```

### PR DESCRIPTION
**Pull request by NiallRees**
Resolves #3578

### Description

Adds logic to `compare` in semver.py to handle pre-release versions.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: test/unit/test_semver.py [hunk 1]
```diff
@@ -134,3 +136,15 @@ def test__resolve_to_specific_version(self):
...
                 ['0.0.3', '0.1.4', '0.0.5']),
             '0.0.5')
+
+        self.assertEqual(
+            resolve_to_specific_version(
+                create_range('=0.4.5a2', '=0.4.5a2'),
+                ['0.4.5a1', '0.4.5a2']),
+            '0.4.5a2')
+
+        self.assertEqual(
```
Comment (NiallRees) [inline test/unit/test_semver.py:146]: Testing the specific case raised by @lbk-fishtown in https://github.com/dbt-labs/dbt/issues/3578

File: test/unit/test_semver.py [hunk 2]
```diff
@@ -134,3 +136,15 @@ def test__resolve_to_specific_version(self):
                 create_range(None, '<=0.0.5'),
                 ['0.0.3', '0.1.4', '0.0.5']),
             '0.0.5')
+
+        self.assertEqual(
+            resolve_to_specific_version(
```
Comment (NiallRees) [inline test/unit/test_semver.py:141]: Testing the specific case raised in https://github.com/dbt-labs/dbt/issues/3609

File: test/unit/test_semver.py [hunk 3]
```diff
@@ -134,3 +136,39 @@ def test__resolve_to_specific_version(self):
...
+                create_range('>=1.0.0', None),
+                ['1.0.0', '1.1.0a1', '1.1.0', '1.2.0a1', '1.2.0']),
+            '1.2.0')
+
+        self.assertEqual(
+            resolve_to_specific_version(
+                create_range('>=1.0.0', '<1.2.0'),
+                ['1.0.0', '1.1.0a1', '1.1.0', '1.2.0a1', '1.2.0']),
+            '1.1.0')
```
Comment (NiallRees) [inline test/unit/test_semver.py:174]: @jtcohen6 does the behaviour of the above four test cases make sense to you?

File: test/unit/test_semver.py [hunk 4]
```diff
@@ -134,3 +136,39 @@ def test__resolve_to_specific_version(self):
...
+            resolve_to_specific_version(
+                create_range('>=1.0.0', None),
+                ['1.0.0', '1.1.0a1', '1.1.0', '1.2.0a1']),
+            '1.2.0a1')
+
+        self.assertEqual(
+            resolve_to_specific_version(
+                create_range('>=1.0.0', '<1.2.0'),
+                ['1.0.0', '1.1.0a1', '1.1.0', '1.2.0a1']),
+            '1.1.0')
```
Comment (NiallRees) [inline test/unit/test_semver.py:162]: Enabled by https://github.com/dbt-labs/dbt/pull/3624/files#diff-b229315589aec2b4aeba481bf11d9e45d550387e96e15440f5737dac43b40df0R134-R139

File: test/unit/test_semver.py (hunk 3 above)
Comment (jtcohen6) [reply test/unit/test_semver.py:174]: Those all look spot on!

**Review by jtcohen6 — APPROVED** [review]
Heroic work @NiallRees !!

Thanks for diving into the semver logic, and big props for doing it on such a quick turnaround. We can get this in for v0.20.1—and I agree that's the right place for this. It isn't a breaking change, except in the sense of changing a behavior that's buggy + unintuitive, and this will grant package maintainers much greater peace of mind going forward.

**Comment by lbk-fishtown** [comment]
Thanks Niall and Jeremy for looking into this one!

File: CHANGELOG.md [hunk 5]
```diff
@@ -32,10 +32,16 @@ Contributors:
...
+### Features
+- Adds `install-prerelease` parameter to hub packages in `packages.yml`. When set to `True`, allows prerelease packages to be installed. By default, this parameter is False unless explicitly set to True.
+
 ### Fixes
 - Fix `store_failures` config when defined as a modifier for `unique` and `not_null` tests ([#3575](https://github.com/fishtown-analytics/dbt/issues/3575), [#3577](https://github.com/fishtown-analytics/dbt/pull/3577))
 - Fix `where` config with `relationships` test by refactoring test SQL. Note: The default `relationships` test now includes CTEs, and may need reimplementing on adapters that don't support CTEs nested inside subqueries. ([#3579](https://github.com/fishtown-analytics/dbt/issues/3579), [#3583](https://github.com/fishtown-analytics/dbt/pull/3583))
+- Fix `dbt deps` version comparison logic which was causing incorrect pre-release package versions to be installed. ([#3578](https://github.com/dbt-labs/dbt/issues/3578), [#3609](https://github.com/dbt-labs/dbt/issues/3609))
 
+Contributors:
+- [@NiallRees](https://github.com/NiallRees) ([#3623](https://github.com/dbt-labs/dbt/pull/3623))
```
Comment (NiallRees) [inline CHANGELOG.md:44]: Not sure what happened here 🙈 https://github.com/dbt-labs/dbt/pull/3639

Comment (jtcohen6) [reply CHANGELOG.md:44]: No worries! This happens all the time...

**Pull request merged**